### PR TITLE
use clang 12

### PR DESCRIPTION
**Pull request by lizan**
Signed-off-by: Lizan Zhou <lizan@tetrate.io>

### CODE REVIEW

**Comment by phlax** [comment]
hi @lizan not sure if you saw my post in the build-tools slack channel - how about using apt.llvm.org to install llvm - iiuc it would reduce the size of this image by ~1GB

not sure if all architectures are supported - maybe we could at least use it where it is

**Review by keith — DISMISSED** [review]

**Comment by lizan** [comment]
@phlax the reason that we're not using apt.llvm.org is the libc++.a from there isn't built with PIC so it won't let us build most statically linked binary as PIE.

**Comment by phlax** [comment]
is there any way we can make our llvm install smaller - iiuc the reason its so big is that llvm itself is statically linked

**Review by phlax — DISMISSED** [review]
lgtm, thanks

**Comment by moderation** [comment]
How about a 40MB binary for LLVM 12?

https://zig.news/kristoff/compile-a-c-c-project-with-zig-368j

**Review by wrowe — APPROVED** [review]
Looks good here, but should we bump to 12.0.1 at this point to pick up any small bug fixes?

**Comment by wrowe** [comment]
@lizan if we can bump this ahead to 12.0.1 @phlax and I would like fresh build images to do further testing on the new toolchain structure assembled by @yanavlasov - and this may clear up an optimization segfault in clang-cl observed in the optimization pass.

File: build_container/build_container_ubuntu.sh [hunk 1]
```diff
@@ -92,19 +92,19 @@ PACKAGES=(
...
 case $ARCH in
     'ppc64le' )
         LLVM_DISTRO=powerpc64le-linux-ubuntu-18.04
-        LLVM_SHA256SUM=acf6bd491d60d126254dc867b2c64c9f1077a4b3dabc0381c31fd8e72aacc4f2
+        LLVM_SHA256SUM=271b9605b74d904d3cc05dd6a61e927fd5a46d5f6b7541cdc67186eb02b22e4c
         ;;
     'x86_64' )
         LLVM_DISTRO=x86_64-linux-gnu-ubuntu-16.04
-        LLVM_SHA256SUM=9694f4df031c614dbe59b8431f94c68631971ad44173eecc1ea1a9e8ee27b2a3
+        LLVM_SHA256SUM=6b3cc55d3ef413be79785c4dc02828ab3bd6b887872b143e3091692fc6acefe7
```
Comment (wrowe) [inline build_container/build_container_ubuntu.sh:103]: With ubuntu 16.04 now expired in April, should this be kicked up to 18.04 LTS?

Comment (lizan) [reply build_container/build_container_ubuntu.sh:103]: That's what LLVM release provides, and it works on 16.04+. our image base are 18.04.

**Comment by wrowe** [comment]
Ok... this is the failure we were seeing for ubuntu, re-running to confirm...

```
clang+llvm-12.0.1-x86_64-linux-gnu-ubuntu-16.04.tar.xz: OK
mv: cannot stat './clang+llvm-12.0.1-x86_64-linux-gnu-ubuntu-16.04': No such file or directory
The command '/bin/sh -c ./build_container_ubuntu.sh' returned a non-zero code: 1
```

**Review by wrowe — APPROVED** [review]
Once I resubmitted, everything is fine. Approving...

**Comment by yanavlasov** [comment]
Once it generates new toolchains I will try it in envoy repo.

**Pull request merged**